lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -507,16 +597,217 @@ bool VerifySignatureMatches(llvm::Module *pModule, DXIL::SignatureKind SigKind,
+          RuntimeInfo1->SigPatchConstOrPrimVectors &&
+          RuntimeInfo1->SigInputVectors) {
+        uint32_t TableSizeInBytes =
+            sizeof(uint32_t) * PSVComputeInputOutputTableDwords(
+                                   RuntimeInfo1->SigInputVectors,
+                                   RuntimeInfo1->SigPatchConstOrPrimVectors);
+        INCREMENT_POS(TableSizeInBytes);
+      }
+
+      if (RuntimeInfo1->ShaderStage == (unsigned)DXIL::ShaderKind::Domain &&
+          RuntimeInfo1->SigOutputVectors[0] &&
+          RuntimeInfo1->SigPatchConstOrPrimVectors) {
+        uint32_t TableSizeInBytes =
+            sizeof(uint32_t) * PSVComputeInputOutputTableDwords(
+                                   RuntimeInfo1->SigPatchConstOrPrimVectors,
+                                   RuntimeInfo1->SigOutputVectors[0]);
+        INCREMENT_POS(TableSizeInBytes);
+      }
+    }
+    IsValid = PSVSize == Offset;
+#undef INCREMENT_POS
+  }
+  bool ValidatePSVInit(PSVInitInfo PSVInfo, ValidationContext &ValCtx) {
+    if (PSVRuntimeInfoSize != PSVInfo.RuntimeInfoSize()) {
+      ValCtx.EmitFormatError(ValidationRule::ContainerContentMatches,
+                             {"PSVRuntimeInfoSize", "PSV0",
+                              std::to_string(PSVRuntimeInfoSize),
+                              std::to_string(PSVInfo.RuntimeInfoSize())});
+      return false;
+    }
+    if (PSVNumResources &&
+        PSVResourceBindInfoSize != PSVInfo.ResourceBindInfoSize()) {
+      ValCtx.EmitFormatError(ValidationRule::ContainerContentMatches,
+                             {"PSVResourceBindInfoSize", "PSV0",
+                              std::to_string(PSVResourceBindInfoSize),
+                              std::to_string(PSVInfo.ResourceBindInfoSize())});
+      return false;
+    }
+    if (RuntimeInfo1 &&
+        (RuntimeInfo1->SigInputElements || RuntimeInfo1->SigOutputElements ||
+         RuntimeInfo1->SigPatchConstOrPrimElements) &&
+        PSVSignatureElementSize != PSVInfo.SignatureElementSize()) {
+      ValCtx.EmitFormatError(ValidationRule::ContainerContentMatches,
+                             {"PSVSignatureElementSize", "PSV0",
+                              std::to_string(PSVSignatureElementSize),
+                              std::to_string(PSVInfo.SignatureElementSize())});
+      return false;
+    }
+    return true;
+  }
+
+private:
+  const char *GetPtrAtOffset(const void *BasePtr, uint32_t Offset) const {
+    return (const char *)BasePtr + Offset;
+  }
+  uint32_t GetUint32AtOffset(const void *BasePtr, uint32_t Offset) const {
+    return *(const uint32_t *)GetPtrAtOffset(BasePtr, Offset);
+  }
+};
+
 static void VerifyPSVMatches(ValidationContext &ValCtx, const void *pPSVData,
                              uint32_t PSVSize) {
+  SimplePSV SimplePSV(pPSVData, PSVSize);
+  if (!SimplePSV.IsValid) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerPartMatches,
+                           {"Pipeline State Validation"});
```

Comment:
So, this is the generic error that the container part must match the module.  If this is invalid though, I think the error could be more specific, indicating that the PSV0 part is malformed.

Reply:
Updated.